test/integration/podgc/podgc_test.go

```diff
@@ -28,152 +28,238 @@ import (
+			if err != nil {
+				t.Fatalf("Failed to create node '%v', err: %v", node.Name, err)
+			}
 
-	// taint the node with the out-of-service taint
-	err = testutils.AddTaintToNode(cs, pod.Spec.NodeName, v1.Taint{Key: v1.TaintNodeOutOfService, Value: "", Effect: v1.TaintEffectNoExecute})
-	if err != nil {
-		t.Fatalf("Failed to taint node: %v, err: %v", pod.Spec.NodeName, err)
-	}
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "testpod",
+					Namespace: testCtx.NS.Name,
+				},
+				Spec: v1.PodSpec{
+					NodeName: node.Name,
+					Containers: []v1.Container{
+						{Name: "foo", Image: "bar"},
+					},
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodRunning,
+				},
+			}
+			if test.withFinalizer {
+				pod.ObjectMeta.Finalizers = []string{"test.k8s.io/finalizer"}
+			}
+
+			pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
+			}
+			if test.withFinalizer {
+				defer testutils.RemovePodFinalizers(testCtx.ClientSet, t, []*v1.Pod{pod})
+			}
 
-	// wait until the pod is deleted
-	err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
-		updatedPod, err := cs.CoreV1().Pods(pod.Namespace).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
-		if err == nil {
-			return updatedPod == nil, nil
-		}
-		// there was an error
-		if apierrors.IsNotFound(err) {
-			return true, nil
-		}
-		return false, err
-	})
-	if err != nil {
-		t.Fatalf("Error '%v' while waiting for the pod '%v' to be deleted", err, klog.KObj(pod))
+			// trigger termination of the pod, but with long grace period so that it is not removed immediately
+			err = cs.CoreV1().Pods(testCtx.NS.Name).Delete(testCtx.Ctx, pod.Name, metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(300)})
+			if err != nil {
+				t.Fatalf("Error: '%v' while deleting pod: '%v'", err, klog.KObj(pod))
+			}
+			// wait until the pod is terminating
+			err = wait.PollImmediate(time.Second, time.Second*15, testutils.PodIsGettingEvicted(cs, pod.Namespace, pod.Name))
+			if err != nil {
+				t.Fatalf("Error '%v' while waiting for the pod '%v' to be terminating", err, klog.KObj(pod))
+			}
+			// taint the node with the out-of-service taint
+			err = testutils.AddTaintToNode(cs, pod.Spec.NodeName, v1.Taint{Key: v1.TaintNodeOutOfService, Value: "", Effect: v1.TaintEffectNoExecute})
+			if err != nil {
+				t.Fatalf("Failed to taint node: %v, err: %v", pod.Spec.NodeName, err)
+			}
+			if test.withFinalizer {
+				// wait until the pod phase is set as expected
+				err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
+					var e error
+					pod, e = cs.CoreV1().Pods(pod.Namespace).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+					if e != nil {
+						return true, e
+					}
+					return test.wantPhase == pod.Status.Phase, nil
+				})
+				if err != nil {
```

Comment:
Check the condition too.

Reply:
I've added a check that the condition is not there. Currently, we append the condition in gc_controller only for orphaned pods, as it seems to be the responsibility of other components to add an appropriate condition in such a case. Maybe, actually, we should also set the phase to Failed to orphaned pods only @alculquicondor ?

Reply:
Oops, right, there is no condition here.

I think we should set the phase to Failed here, because the pod is being fully removed regardless.